signalfx-async-wrapper.js

```diff
@@ -0,0 +1,92 @@
+'use strict';
+
+const sfxHelper = require('./signalfx-helper');
+
+let coldStart = true;
+
+class SignalFxWrapper {
+  constructor(
+    originalObj,
+    originalFn,
+    originalEvent,
+    originalContext,
+    originalCallback,
+    dimensions,
+    accessToken
+  ) {
+    this.originalObj = originalObj;
+    this.originalFn = originalFn;
+    this.originalEvent = originalEvent;
+    this.originalContext = originalContext;
+    this.originalCallback = originalCallback;
+
+    sfxHelper.setAccessToken(accessToken);
+    sfxHelper.setLambdaFunctionContext(this.originalContext, dimensions);
+    sfxHelper.sendCounter('function.invocations', 1);
+    if (coldStart) {
+      sfxHelper.sendCounter('function.cold_starts', 1);
+      coldStart = false;
+    }
+    return this;
+  }
+
+  async invoke() {
+    const startTime = new Date().getTime();
+    let sent;
+    const sendDuration = async () => {
+      if (sent) {
+        return;
+      }
+
+      sfxHelper.sendGauge('function.duration', new Date().getTime() - startTime);
+      sent = true;
+
+      await sfxHelper.waitForAllSends();
+    }
+
+    const callbackWrapper = (err, result) => {
+      sendDuration().then(() => {
+        this.originalCallback(err, result);
+      })
+    };
+
+    let result;
+    try {
+      result = await this.originalFn.call(
+        this.originalObj,
+        this.originalEvent,
+        this.originalContext,
+        callbackWrapper
+      );
+
+      await sendDuration();
+    } catch (err) {
+      sfxHelper.sendCounter('function.errors', 1);
+      await sendDuration();
+
+      // should just return js exception instead of doing additional processing: https://aws.amazon.com/blogs/compute/node-js-8-10-runtime-now-available-in-aws-lambda/
+      return err;
```

Comment:
In the non-async wrapper, we call the original callback from inside the catch if it hasn't been called yet, after sending the duration metric. Are we missing that here

Reply:
Returning an error will do the job. Like you suggested, we are exposing the callback to the user in case they still want to use it, however, it is not a part of user driven workflow when there's an exception so I doubt that the callback needs to be called.

Reply:
Please see the example here: https://aws.amazon.com/blogs/compute/node-js-8-10-runtime-now-available-in-aws-lambda/